File: src/B_2178/B_2178.java
package B_2178;

import java.io.*;
import java.util.*;


class pos{
    int x;
    int y;
    public pos(int x, int y) {
        this.x = x;
        this.y = y;
    }
}
public class B_2178 {
    static int N,M;
    static int[][] map;
    static boolean[][] visited;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        map = new int[N][M];
        visited = new boolean[N][M];
        for(int i=0; i<N; i++){
            String maze = br.readLine();
            for(int k=0; k<M; k++){
                map[i][k] = maze.charAt(k)-'0';
            }
        }
        visited[0][0]=true;
        bfs(0,0);
        System.out.print(map[N-1][M-1]);
    }
    static void bfs(int x, int y){
        int[] dx = {1,-1,0,0};
        int[] dy = {0,0,1,-1};
        Queue<pos> q = new LinkedList<>();
        q.add(new pos(x,y));
        while(!q.isEmpty()){
            pos P = q.poll();
            for(int i=0; i<4; i++){
                int NextX = P.x+dx[i];
                int NextY = P.y+dy[i];
                if(NextX < 0 || NextX >=N || NextY < 0 || NextY >=M) continue;
                if(visited[NextX][NextY] || map[NextX][NextY]==0) continue;
                visited[NextX][NextY]=true;
                map[NextX][NextY] = map[P.x][P.y]+1;
                q.add(new pos(NextX,NextY));
            }
        }
    }

}

File: src/B_2178/B_2178.kt
package B_2178

import java.io.*
import java.util.*
import kotlin.properties.Delegates

data class Dot(var x:Int , var y:Int)
var map:Array<IntArray> = arrayOf()
var visited:Array<BooleanArray> = arrayOf()
var N by Delegates.notNull<Int>()
var M by Delegates.notNull<Int>()
fun main(args: Array<String>)=with(BufferedReader(InputStreamReader(System.`in`))){
    var st = StringTokenizer(readLine()," ")
    N = st.nextToken().toInt()
    M = st.nextToken().toInt()
    map = Array<IntArray>(N){IntArray(M)}
    visited = Array<BooleanArray>(N){BooleanArray(M) {false} }
    for(i in 0 until N){
        var maze = readLine()
        for(k in 0 until M){
            map[i][k] = maze[k]-'0'
        }
    }
    visited[0][0]=false
    bfs(0,0)
    print(map[N-1][M-1])
}
fun bfs(x:Int, y:Int){
    val dx = intArrayOf(1,-1,0,0)
    val dy = intArrayOf(0,0,1,-1)
    var q:Queue<Dot> = LinkedList()
    q.add(Dot(x,y))
    while(q.isNotEmpty()){
        var d:Dot = q.poll()
        for(i in 0 until 4){
            var NX = d.x + dx[i]
            var NY = d.y + dy[i]
            if(NX < 0 || NY < 0 || NX >= N || NY >= M) continue
            if(visited[NX][NY] || map[NX][NY] == 0) continue
            q.add(Dot(NX,NY))
            map[NX][NY] = map[d.x][d.y]+1
            visited[NX][NY]=true
        }
    }
}